File: src/main/java/de/blazemcworld/fireflow/code/node/impl/event/player/OnPlayerChatNode.java
package de.blazemcworld.fireflow.code.node.impl.event.player;

import de.blazemcworld.fireflow.code.CodeEvaluator;
import de.blazemcworld.fireflow.code.CodeThread;
import de.blazemcworld.fireflow.code.node.Node;
import de.blazemcworld.fireflow.code.type.PlayerType;
import de.blazemcworld.fireflow.code.type.SignalType;
import de.blazemcworld.fireflow.code.type.StringType;
import de.blazemcworld.fireflow.code.value.PlayerValue;
import net.minestom.server.event.player.PlayerChatEvent;

public class OnPlayerChatNode extends Node {
    
    private final Output<Void> signal;
    private final Output<PlayerValue> player;
    private final Output<String> message;

    public OnPlayerChatNode() {
        super("on_player_chat");

        signal = new Output<>("signal", SignalType.INSTANCE);
        player = new Output<>("player", PlayerType.INSTANCE);
        message = new Output<>("message", StringType.INSTANCE);
        player.valueFromThread();
        message.valueFromThread();
    }

    @Override
    public void init(CodeEvaluator evaluator) {
        evaluator.events.addListener(PlayerChatEvent.class, event -> {
            CodeThread thread = evaluator.newCodeThread();
            thread.setThreadValue(player, new PlayerValue(event.getPlayer()));
            thread.setThreadValue(message, event.getMessage());
            thread.sendSignal(signal);
            thread.clearQueue();
        });
    }

    @Override
    public Node copy() {
        return new OnPlayerChatNode();
    }

}

File: src/main/kotlin/de/blazemcworld/fireflow/node/impl/player/event/OnPlayerChatNode.kt
package de.blazemcworld.fireflow.node.impl.player.event

import de.blazemcworld.fireflow.node.*
import net.minestom.server.event.EventFilter
import net.minestom.server.event.EventNode
import net.minestom.server.event.player.PlayerChatEvent
import net.minestom.server.item.Material

object OnPlayerChatNode : BaseNode("On Player Chat", Material.PLAYER_HEAD) {
    private val signal = output("Signal", SignalType)
    private val player = output("Player", PlayerType)
    private val message = output("Message", TextType)

    override fun setup(ctx: NodeContext) {
        val node = EventNode.type("chat", EventFilter.INSTANCE)
        val events = ctx.global.space.playInstance.eventNode().addChild(node)

        node.addListener(PlayerChatEvent::class.java) {
            val evalCtx = EvaluationContext(ctx.global)
            evalCtx[ctx[player]] = { PlayerReference(it.player, ctx.global.space) }
            evalCtx[ctx[message]] = { it.message }
            evalCtx.emit(ctx[signal], now=true)
        }

        ctx.global.onDestroy += {
            events.removeChild(node)
        }
    }

}